IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/Browser.cs

```diff
@@ -0,0 +1,281 @@
+        /// <param name="by">Selenium "By" statement to find the element</param>
+        /// <param name="secondsToWait">Seconds to wait while retrying before failing</param>
+        /// <returns></returns>
+        public ReadOnlyCollection<WebElement> FindElements(By by, int secondsToWait = 5)
+        {
+            Console.WriteLine($"Attempting to find all elements using selector: {by}");
+
+            WebDriverWait wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(secondsToWait));
+            wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.ElementExists(by));
+
+            return new ReadOnlyCollection<WebElement>(
+                            Driver.FindElements(by)
+                                    .Select(webElement => new WebElement(by, Driver))
+                                    .ToList());
+        }
+
+        /// <summary>
+        ///Waits until a function evaluates to true OR times out after a specified period of time
+        /// </summary>
+        /// <param name="func">Function to evaluate</param>
+        /// <param name="secondsToWait">Secondes to wait until timeout / return false</param>
+        /// <returns></returns>
+        public bool WaitFor(Func<bool> func, int secondsToWait = 15)
+        {
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+            DateTime end = DateTime.Now.AddSeconds(secondsToWait);
+            do
+            {
+                // Do a brief wait first, on the assumption that C# hasn't evaluated the delegate yet
+                // This is to test to see if Selenium is tying up processes on the webpage under test
+                Task.Delay(500).Wait();
+                try
+                {
+                    if (func())
+                    {
+                        return true;
+                    }
+                }
+                catch (WebDriverTimeoutException)
+                {
+                    return false;
+                }
+
+            } while (DateTime.Now <= end);
+            return false;
+        }
+
+        public bool IsElementDisplayedAndExisting(By by, int secondsToWait = 15)
+        {
+            // Now that we have a WebElement.Initialized property, see if it makes sense to refactor this logic to look at that instead?
+            // Or is it worth still re-trying, just in case the site was in a bad state prior to calling this?
+            // If the latter, the default secondsToWait can be kicked down to maybe 5.
+            WebDriverWait wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(secondsToWait));
+            try
+            {
+                wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.ElementExists(by));
+            }
+            catch (WebDriverTimeoutException) { }
+            IReadOnlyCollection<IWebElement> elements = Driver.FindElements(by);
+            if (elements.Count > 0)
+            {
+                return elements.ElementAt(0).Displayed;
```

Comment:
Rather than `ElementAt(0)` you probably want `.First()`

Reply:
Fixed